### PR TITLE
Make atexit callbacks threadsafe

### LINKED ISSUE #923: atexit callbacks need to be threadsafe
Apparently POSIX doesn't specify whether atexit callbacks are called before or after threads are terminated during process shutdown.  Following suggests that in fact, they are called while threads still running!

  https://stackoverflow.com/questions/39655868/what-does-the-posix-standard-say-about-thread-stacks-in-atexit-handlers-what

We should make our atexit callbacks threadsafe, simply by using cmpexch to clean things up.

### PR DESCRIPTION
**Pull request by ebraminio**
Fixes #923 hopefully, wanted to go for some macro but felt better to keep it simple, let me know if is wrong or anything.

### CODE REVIEW

File: src/hb-common.cc [hunk 1]
```diff
@@ -60,12 +60,12 @@ _hb_options_init (void)
 
 /**
  * hb_tag_from_string:
- * @str: (array length=len) (element-type uint8_t): 
- * @len:
```
Comment (behdad) [inline src/hb-common.cc]: Why all the whitespace noise? Or do you have an editor that automatically does these?

**Comment by behdad** [comment]
No this is not how you do it.  It's not that multiple copies of the atexit can run. It's that it can race with other code using these. Let me fix one and you see what's needed. No new variables are needed.

**Comment by behdad** [comment]
See this: https://github.com/harfbuzz/harfbuzz/commit/5aa2c6e194bc75637d3f4862a096c0e7100072e2

File: src/hb-common.cc (hunk 1 above)
Comment (ebraminio) [reply src/hb-common.cc]: Yes, that is the case with the our current `.editorconfig` which requests it.

File: src/hb-shape.cc [hunk 2]
```diff
@@ -51,7 +51,12 @@ static const char **static_shaper_list;
 static
 void free_static_shaper_list (void)
 {
-  free (static_shaper_list);
+retry:
+  char **shaper_list = (char **) hb_atomic_ptr_get (&static_shaper_list);
+  if (!hb_atomic_ptr_cmpexch (static_shaper_list, *shaper_list, nullptr))
+    goto retry;
```
Comment (ebraminio) [inline src/hb-shape.cc]: Hmm, this one doesn't seem right.

**Comment by ebraminio** [comment]
PTAL

**Comment by behdad** [comment]
Thanks. Looks good.

It occurred to me that this is rather futile. If we free these while another thread might use them, there's an unfixable race condition. I'm happy pushing this in though.

**Comment by ebraminio** [comment]
> It occurred to me that this is rather futile. If we free these while another thread might use them, there's an unfixable race condition.

I wanted to ask about the benefit on that situation also.

**Pull request merged**